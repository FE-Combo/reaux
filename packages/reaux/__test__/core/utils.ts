import {setStateAction, setHelperLoadingAction, setHelperLanguageAction, setHelperExceptionAction} from "../../src/core/utils";
import {Exception} from "../../src/type";

test("Set State Action", () => {
    expect(setStateAction("moduleName", {name: "name", age: 11})).toEqual({
        type: "@@framework/setState",
        payload: {module: "moduleName", state: {name: "name", age: 11}},
    });

    expect(setStateAction("moduleName2", {name: "name2", age: 22})).toEqual({
        type: "@@framework/setState",
        payload: {module: "moduleName2", state: {name: "name2", age: 22}},
    });
});

test("Set Loading In Helper", () => {
    expect(setHelperLoadingAction("identifier", true)).toEqual({
        type: "@@framework/setHelper/loading",
        payload: {identifier: "identifier", hasShow: true},
    });

    expect(setHelperLoadingAction("identifier2", false)).toEqual({
        type: "@@framework/setHelper/loading",
        payload: {identifier: "identifier2", hasShow: false},
    });
});

test("Set Language In Helper", () => {
    expect(setHelperLanguageAction("CN")).toEqual({
        type: "@@framework/setHelper/language",
        payload: "CN",
    });

    expect(setHelperLanguageAction("EN")).toEqual({
        type: "@@framework/setHelper/language",
        payload: "EN",
    });
});

test("Set Exception In Helper", () => {
    expect(setHelperExceptionAction({} as Exception)).toEqual({
        type: "@@framework/setHelper/exception",
        payload: {},
    });
});
