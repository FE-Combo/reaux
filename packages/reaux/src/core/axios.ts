import axios, {AxiosRequestConfig, Method} from "axios";

export function ajax<Request, Response>(method: Method, path: string, pathParams: object, request: Request): Promise<Response> {
    const config: AxiosRequestConfig = {method, url: url(path, pathParams)};

    if (method === "GET" || method === "DELETE") {
        config.params = request;
    } else if (method === "POST" || method === "PUT" || method === "PATCH") {
        config.data = request;
    }

    return axios.request(config).then(response => response.data);
}

function url(pattern: string, params: object): string {
    if (!params) {
        return pattern;
    }
    let url = pattern;
    Object.entries(params).forEach(([name, value]) => {
        const encodedValue = encodeURIComponent(value.toString());
        url = url.replace(":" + name, encodedValue);
    });
    return url;
}
