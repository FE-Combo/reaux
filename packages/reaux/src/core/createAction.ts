import {ActionType, ModelLifeCycle, ExceptionHandler} from "../type";

type ActionCreator<H> = H extends (...args: infer P) => any ? ((...args: P) => ActionType<P>) : never;
type HandlerKeys<H> = {[K in keyof H]: H[K] extends (...args: any[]) => any ? K : never}[Exclude<keyof H, keyof ModelLifeCycle | keyof ExceptionHandler>];
type ActionCreators<H> = {readonly [K in HandlerKeys<H>]: ActionCreator<H[K]>};

export interface ActionsType {
    [key: string]: (...payload: any[]) => ActionType<any[]>;
}

interface ActionHandlers {
    [key: string]: (...args: any[]) => any;
}

export default function createAction<H extends object & {moduleName: string}>(handler: H) {
    const moduleName = handler.moduleName;
    const keys = getPrototypeOfExceptConstructor(handler);
    const actions = {} as ActionCreators<H>;
    const actionsHandler = {} as ActionHandlers;
    keys.forEach(actionType => {
        const method = handler[actionType];
        const qualifiedActionType = `${moduleName}/${actionType}`;
        actions[actionType] = (...payload: any[]) => ({type: qualifiedActionType, payload});
        actionsHandler[qualifiedActionType] = method.bind(handler);
    });

    return {actions, actionsHandler};
}

function getPrototypeOfExceptConstructor(object: object): string[] {
    return Object.getOwnPropertyNames(Object.getPrototypeOf(object)).filter(key => key !== "constructor");
}
