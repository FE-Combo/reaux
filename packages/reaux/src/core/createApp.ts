import {AppView} from "../type";

export default function createApp<T extends AppView>(callback?: (app: AppView) => T): T {
    const app = {
        actionHandler: {},
        modules: {},
        // TODO: Why use object ?????????
        exceptionHandler: {},
    } as T;
    return callback ? callback(app) : app;
}
