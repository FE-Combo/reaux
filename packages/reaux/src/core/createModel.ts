/*
  proxy store
*/

import {StateView} from "../type";

abstract class ModelProperty<State = {}> {
    abstract readonly moduleName: string;
    abstract readonly initState: State;
    abstract state: Readonly<State>;
    abstract rootState: Readonly<StateView>;
    abstract setState(newState: Partial<State>): void;
}

abstract class ModelLifeCycle<R = any> {
    abstract onReady(): R;
    abstract onLoad(): R;
    abstract onUnload(): R;
    abstract onHide(): R;
}

type DispatchState<State> = (module: string, state: Partial<State>, type: string) => any;

export default function createModel<State>(allState: StateView, dispatchState: DispatchState<State>) {
    return class Model extends ModelProperty<State> {
        public constructor(readonly moduleName: string, readonly initState: State) {
            super();
            dispatchState(moduleName, initState, `@@${moduleName}/initState`);
        }

        get state(): Readonly<State> {
            return allState.app[this.moduleName];
        }

        get rootState(): Readonly<StateView> {
            return allState;
        }

        setState(newState: Partial<State>) {
            dispatchState(this.moduleName, newState, `@@${this.moduleName}/setState[${Object.keys(newState).join(",")}]`);
        }
    };
}

export function createBaseModelPromise<State>(allState: StateView, dispatchState: DispatchState<State>) {
    const BaseModel = createModel(allState, dispatchState);
    return class Model extends BaseModel implements ModelLifeCycle {
        async onReady() {
            // extends to be overrode
        }

        async onLoad() {
            // extends to be overrode
        }

        async onUnload() {
            // extends to be overrode
        }

        async onHide() {
            // extends to be overrode
        }
    };
}

export function createBaseModelGenerator<State>(allState: StateView, dispatchState: DispatchState<State>) {
    const BaseModel = createModel(allState, dispatchState);
    return class Model extends BaseModel implements ModelLifeCycle {
        *onReady(): any {
            // extends to be overrode
        }

        *onLoad(): any {
            // extends to be overrode
        }

        *onUnload(): any {
            // extends to be overrode
        }

        *onHide(): any {
            // extends to be overrode
        }
    };
}
