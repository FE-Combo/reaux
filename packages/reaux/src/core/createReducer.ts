import {Reducer, combineReducers, ReducersMapObject} from "redux";
import {StateView, LoadingState, ActionType, ActionPayload, LoadingActionType} from "../type";

export const SET_STATE_ACTION: string = "@@framework/setState";
export const LOADING_ACTION: string = "@@framework/loading";

export function setStateAction<State>(module: string, state: Partial<State>, type: string): ActionType<ActionPayload> {
    return {
        type,
        name: SET_STATE_ACTION,
        payload: {module, state},
    };
}

export function setLoadingAction(identifier: string, hasShow: boolean): ActionType<LoadingActionType> {
    return {
        type: LOADING_ACTION,
        payload: {identifier, hasShow},
    };
}

function appReducer(state: StateView["app"] = {}, action: ActionType<any>): StateView["app"] {
    if (action.name === SET_STATE_ACTION) {
        const {module, state: moduleState} = action.payload as ActionPayload;
        return {...state, [module]: {...state[module], ...moduleState}};
    }
    return state;
}

function loadingReducer(state: LoadingState = {}, action: ActionType<any>): LoadingState {
    if (action.type === LOADING_ACTION) {
        const payload = action.payload as LoadingActionType;
        const count = state[payload.identifier] || 0;
        return {
            ...state,
            [payload.identifier]: count + (payload.hasShow ? 1 : -1),
        };
    }
    return state;
}

export default function createReducer<T extends StateView>(callback?: (result: ReducersMapObject<StateView, any>) => ReducersMapObject<any, any>): Reducer<T> {
    const reducers: ReducersMapObject<StateView, any> = {app: appReducer, loading: loadingReducer};
    const resultReducers = callback ? callback(reducers) : reducers;
    return combineReducers(resultReducers);
}
