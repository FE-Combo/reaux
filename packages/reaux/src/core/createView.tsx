/*
    proxy LifeCycle
*/

import React from "react";
import {BaseModel} from "../type";

export function createCView<H extends BaseModel<{}, any>>(handler: H, Component: React.ComponentType<any>) {
    return class View<P extends {} = {}> extends React.PureComponent<P> {
        constructor(props: P) {
            super(props);
            handler.onReady();
        }

        componentDidMount() {
            handler.onLoad();
        }

        componentDidUpdate() {
            handler.onLoad();
        }

        componentWillUnmount() {
            handler.onUnload();
        }

        render() {
            return <Component {...this.props} />;
        }
    };
}

export function createFCView<H extends BaseModel<{}, any>>(handler: H, Component: React.ComponentType<any>) {
    // TODO:
}
