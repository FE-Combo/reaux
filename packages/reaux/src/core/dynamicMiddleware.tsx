import {compose} from "redux";

// TODO: remove any
const createDynamicMiddleware = () => {
    let allDynamicMiddleware: any[] = [];
    let allAppliedDynamicMiddleware: any[] = [];
    let store: any;

    const enhancer = (_store: any) => {
        store = _store;
        return (next: any) => (action: any) => {
            return (compose(...allAppliedDynamicMiddleware)(next) as any)(action);
        };
    };

    const addMiddleware = (...middleware: any[]) => {
        allAppliedDynamicMiddleware.push(...middleware.map(middleware => middleware(store)));
        allDynamicMiddleware.push(...middleware);
    };

    const removeMiddleware = (middleware: any) => {
        const index = allDynamicMiddleware.findIndex(d => d === middleware);

        if (index === -1) {
            // eslint-disable-next-line no-console
            console.error("Middleware does not exist!", middleware);
            return;
        }

        allDynamicMiddleware = allDynamicMiddleware.filter((_, mdwIndex) => mdwIndex !== index);
        allAppliedDynamicMiddleware = allAppliedDynamicMiddleware.filter((_, mdwIndex) => mdwIndex !== index);
    };

    const resetMiddleware = () => {
        allAppliedDynamicMiddleware = [];
        allDynamicMiddleware = [];
    };

    return {
        enhancer,
        addMiddleware,
        removeMiddleware,
        resetMiddleware,
    };
};

const dynamicMiddlewareInstance = createDynamicMiddleware();

const dynamicMiddleware = dynamicMiddlewareInstance.enhancer;

export const {addMiddleware, removeMiddleware, resetMiddleware} = dynamicMiddlewareInstance;

export {createDynamicMiddleware, dynamicMiddleware};
