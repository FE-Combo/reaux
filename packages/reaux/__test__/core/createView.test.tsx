import React from "react";
import {createCView} from "../../src/core/createView";
import renderer from "react-test-renderer";

test("Create Component View", () => {
    class Model {
        moduleName: string;
        initState: any;
        state: any;
        rootState: any;
        setState: any;
        constructor(moduleName: string, state: any) {
            this.moduleName = moduleName;
            this.state = state;
        }
        onReady() {}
        onLoad() {}
        onUnload() {}
        onHide() {}
    }
    const handler = new Model("module", {});
    class DIV extends React.PureComponent {
        render() {
            return <div>text</div>;
        }
    }
    class Component extends React.PureComponent<any, any> {
        constructor(props: any) {
            super(props);
            this.state = {toggle: false};
        }

        onToggle = () => this.setState({toggle: !this.state.toggle});

        render() {
            return (
                <div onClick={this.onToggle}>
                    {this.state.toggle ? "toggle component" : "component"}
                    <DIV />
                </div>
            );
        }
    }
    const NewComponent = createCView(handler, Component);
    /* 
        Error: [test-renderer] Cannot read property `current` of undefined
        ref: https://github.com/facebook/react/issues/16323
        react and react-test-renderer be same version
     */
    const newComponent = renderer.create(<NewComponent name="name" age={18} />);
    let tree = newComponent.toJSON()!;
    const newDIV = renderer.create(<DIV />);
    let divTree = newDIV.toJSON()!;
    expect(tree).toMatchSnapshot();
    expect(tree.type).toBe("div");
    expect(tree.children).toEqual(["component", divTree]);
    // TODO: Jest error
    // expect(tree.props).toEqual({name: "name", age: 18});

    // manually trigger the callback
    tree.props.onClick();
    // re-rendering
    tree = newComponent.toJSON()!;
    expect(tree).toMatchSnapshot();

    newComponent.update(<NewComponent />);
    newComponent.unmount();
});
