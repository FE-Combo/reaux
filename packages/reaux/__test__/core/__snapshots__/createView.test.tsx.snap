// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Component View 1`] = `
<div
  onClick={[Function]}
>
  component
  <div>
    text
  </div>
</div>
`;

exports[`Create Component View 2`] = `
<div
  onClick={[Function]}
>
  toggle component
  <div>
    text
  </div>
</div>
`;
